### PR TITLE
Ensure successful registration before key refresh

### PR DESCRIPTION
**Pull request by aitorvs**
Task/Issue URL: https://app.asana.com/0/0/1205631225739207/f

### Description
Improve re-key so that:
* it happens once a day
* just re-key when the screen is off
* avoids re-keying attempts when device is idle

### Steps to test this PR

_Test force re-key_
- [x] build internal from this branch
- [x] use force rekey in the netp dev settings
- [x] new private key should be registered

_Test rekey_
- apply the following patch
```diff
Index: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyer.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyer.kt b/network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyer.kt
--- a/network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyer.kt	(revision e44ddfce33069ae2ee2227bddffb24cc89049381)
+++ b/network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyer.kt	(date 1698157737720)
@@ -68,7 +68,7 @@
         val forceOrFalseInProductionBuilds = forceRekey.getAndResetValue()
 
         val millisSinceLastKeyUpdate = System.currentTimeMillis() - networkProtectionRepository.lastPrivateKeyUpdateTimeInMillis
-        if (!forceOrFalseInProductionBuilds && millisSinceLastKeyUpdate < TimeUnit.DAYS.toMillis(1)) {
+        if (!forceOrFalseInProductionBuilds && millisSinceLastKeyUpdate < TimeUnit.SECONDS.toMillis(5)) {
             logcat { "Less than 24h passed, skip re-keying" }
             return
         }
Index: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyScheduler.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyScheduler.kt b/network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyScheduler.kt
--- a/network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyScheduler.kt	(revision e44ddfce33069ae2ee2227bddffb24cc89049381)
+++ b/network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyScheduler.kt	(date 1698157737723)
@@ -42,7 +42,7 @@
         job += coroutineScope.launch(dispatcherProvider.io()) {
             while (isActive && vpnFeaturesRegistry.isFeatureRegistered(NetPVpnFeature.NETP_VPN)) {
                 logcat { "Start periodic re-keying attempts" }
-                delay(TimeUnit.HOURS.toMillis(1)) // try to re-key every 1h
+                delay(TimeUnit.SECONDS.toMillis(10)) // try to re-key every 1h
 
                 netPRekeyer.doRekey()
             }
```
- build internal build
- launch app and enable NetP
- [x] with mobile device screen ON, see `Device not locked, skip re-keying` logcat appear every 10 seconds or so
- [x] with mobile device screen OFF, see `Restarting VPN after clearing client keys` logcat appear every 10 seconds or so

### CODE REVIEW

**Comment by aitorvs** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3712** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3712" target="_blank">Graphite</a> 
    * **PR #3736** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3736" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/3736?utm_source=stack-comment).

File: anvil/anvil-compiler/src/main/java/com/duckduckgo/anvil/compiler/ContributesServiceApiCodeGenerator.kt [hunk 1]
```diff
@@ -95,6 +96,12 @@ class ContributesServiceApiCodeGenerator : CodeGenerator {
                     .addFunction(
                         FunSpec.builder("provides${serviceClassName.simpleName}")
                             .addAnnotation(Provides::class)
+                            .apply {
+                                // Add qualifier annotations as well
+                                qualifierAnnotations.forEach { qualifier ->
+                                    addAnnotation(qualifier.toAnnotationSpec())
+                                }
```
Comment (aitorvs) [inline anvil/anvil-compiler/src/main/java/com/duckduckgo/anvil/compiler/ContributesServiceApiCodeGenerator.kt:103]: Adding ability to use qualifiers with the `@ContributesServiceApi` annotation

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgVpnControllerService.kt [hunk 2]
```diff
@@ -18,12 +18,85 @@ package com.duckduckgo.networkprotection.impl.configuration
...
+                socketProtector.get().protect(sslSocket)
+                return sslSocket
+            }
+        }
+    }
+
+    @Provides
+    @UnprotectedVpnControllerService
+    @SingleInstanceIn(VpnScope::class)
+    fun providesWgVpnControllerService(
```
Comment (aitorvs) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgVpnControllerService.kt:67]: The `UnprotectedVpnControllerService` is a retrofit API for the controller API that will never to through the VPN. 

In fact, we could totally remove the `ProtectedVpnControllerService` as currently there is no use case for calling the controller API through the VPN tun interface

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyScheduler.kt [hunk 3]
```diff
@@ -16,47 +16,35 @@
...
-                    .build()
-                val workerRequest = PeriodicWorkRequestBuilder<NetPRekeyWorker>(1, HOURS)
-                    .boundToVpnProcess(appBuildConfig.applicationId) // this worker is executed in the :vpn process
-                    .setConstraints(constraints)
-                    .build()
-                // Once work is already scheduled, we don't update the interval
-                // since onVpnStarted will be called many times when restarting/reconnecting the VPNService.
-                workManager.enqueueUniquePeriodicWork(DAILY_NETP_REKEY_TAG, ExistingPeriodicWorkPolicy.KEEP, workerRequest)
+        job += coroutineScope.launch(dispatcherProvider.io()) {
+            while (isActive && vpnFeaturesRegistry.isFeatureRegistered(NetPVpnFeature.NETP_VPN)) {
```
Comment (aitorvs) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyScheduler.kt:43]: Ensure we never try to rekey when device is idle.

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyer.kt [hunk 4]
```diff
@@ -18,58 +18,99 @@ package com.duckduckgo.networkprotection.impl.rekey
...
     }
 
+    suspend fun forceRekey() {
+        if (appBuildConfig.isInternalBuild()) {
+            forceRekey.set(true)
+            doRekey()
+        } else {
+            logcat(LogPriority.ERROR) { "Force re-key not allowed in production builds" }
+        }
+    }
```
Comment (aitorvs) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyer.kt:112]: This is to keep the `force rekey` internal build functionality

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/waitlist/NetPWaitlistManager.kt [hunk 5]
```diff
@@ -52,7 +53,7 @@ interface NetPWaitlistManager {
 @SingleInstanceIn(AppScope::class)
 @ContributesBinding(AppScope::class)
 class RealNetPWaitlistManager @Inject constructor(
-    private val service: WgVpnControllerService,
+    @ProtectedVpnControllerService private val service: WgVpnControllerService,
```
Comment (aitorvs) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/waitlist/NetPWaitlistManager.kt:56]: Ditto, ie. this probably can also rather be `@UnprotectedVpnControllerService` at which point we delete the `@ProtectedVpnControllerService` altogether

File: network-protection/network-protection-internal/src/main/java/com/duckduckgo/networkprotection/internal/feature/NetPInternalSettingsActivity.kt [hunk 6]
```diff
@@ -215,7 +213,7 @@ class NetPInternalSettingsActivity : DuckDuckGoActivity() {
 
         binding.forceRekey.setClickListener {
             lifecycleScope.launch {
-                netPRekeyer.doRekey()
+                sendBroadcast(Intent(DebugRekeyReceiver.ACTION_FORCE_REKEY))
```
Comment (aitorvs) [inline network-protection/network-protection-internal/src/main/java/com/duckduckgo/networkprotection/internal/feature/NetPInternalSettingsActivity.kt:216]: We use a broadcast as an IPC mechanism, as rekeying needs to be received/executed by the `:vpn` process

File: network-protection/network-protection-internal/src/main/java/com/duckduckgo/networkprotection/internal/network/WgServerInternalProvider.kt [hunk 7]
```diff
@@ -32,7 +33,7 @@ import javax.inject.Inject
 )
 class WgServerInternalProvider @Inject constructor(
     private val netPServerRepository: NetPServerRepository,
-    private val controllerService: WgVpnControllerService,
+    @ProtectedVpnControllerService private val controllerService: WgVpnControllerService,
```
Comment (aitorvs) [inline network-protection/network-protection-internal/src/main/java/com/duckduckgo/networkprotection/internal/network/WgServerInternalProvider.kt:36]: Ditto

File: network-protection/network-protection-internal/src/main/java/com/duckduckgo/networkprotection/internal/rekey/DebugRekeyReceiver.kt [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+@InjectWith(VpnScope::class)
+@ContributesMultibinding(
+    scope = VpnScope::class,
+    boundType = VpnServiceCallbacks::class,
+)
+class DebugRekeyReceiver @Inject constructor(
+    private val context: Context,
+    private val rekeyer: RealNetPRekeyer,
+    @ProcessName private val processName: String,
```
Comment (aitorvs) [inline network-protection/network-protection-internal/src/main/java/com/duckduckgo/networkprotection/internal/rekey/DebugRekeyReceiver.kt:47]: This is just for logging

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgVpnControllerService.kt [hunk 9]
```diff
@@ -16,14 +16,87 @@
...
+
+        // insertProviderAt trick to avoid error during handshakes
+        Security.insertProviderAt(Conscrypt.newProvider(), 1)
+
+        return customRetrofit.create(WgVpnControllerService::class.java)
+    }
+}
+
 @ContributesServiceApi(AppScope::class)
+@ProtectedVpnControllerService
```
Comment (karlenDimla) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgVpnControllerService.kt:99]: Checked this and you are right this doesn’t seem to be necessary.

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyScheduler.kt (hunk 3 above)
Comment (karlenDimla) [reply network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyScheduler.kt:43]: Question: What will happen to the previously scheduled job DAILY_NETP_REKEY_TAG after this change gets to the client? Will they automatically get cancelled?

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyer.kt [hunk 10]
```diff
@@ -18,59 +18,118 @@ package com.duckduckgo.networkprotection.impl.rekey
...
+@ContributesTo(VpnScope::class)
+object DeviceLockedCheckerModule {
+
+    @Provides
+    @InternalApi
+    fun provideDeviceLockChecker(context: Context): DeviceLockedChecker {
+        val keyguardManager = runCatching { context.getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager }.getOrNull()
+
+        return {
+            keyguardManager?.isDeviceLocked ?: true
```
Comment (karlenDimla) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyer.kt:132]: “```
Returns whether the device is currently locked and requires a PIN, pattern or password to unlock.
Returns:
true if unlocking the device currently requires a PIN, pattern or password.
```”

Wouldn’t this return `false` if the screen is just closed and the user doesn’t have a lock screen?

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgVpnControllerService.kt (hunk 9 above)
Comment (aitorvs) [reply network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgVpnControllerService.kt:99]: As discussed offline, agreed to remove it

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyer.kt [hunk 11]
```diff
@@ -18,59 +18,118 @@ package com.duckduckgo.networkprotection.impl.rekey
...
+        if (!forceOrFalseInProductionBuilds && millisSinceLastKeyUpdate < TimeUnit.DAYS.toMillis(1)) {
             logcat { "Less than 24h passed, skip re-keying" }
             return
         }
-        if (keyguardManager.deviceLocked()) {
-            networkProtectionRepository.privateKey = null
+
+        val newKey = runCatching {
+            val newKeys = KeyPair()
+            wgVpnControllerService.registerKey(RegisterKeyBody(newKeys.publicKey.toBase64(), "*"))
```
[the remark this answers is not in the document]
Comment (aitorvs) [reply network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyer.kt:79]: Good catch! will call `wgServerApi` instead

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyer.kt (hunk 10 above)
Comment (aitorvs) [reply network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyer.kt:132]: I will change the check to `(keyguardManager?.isDeviceLocked == true || powerManager?.isInteractive == false)`

**Review by karlenDimla — APPROVED** [review]

**Review by karlenDimla — APPROVED** [review]

**Pull request merged**